fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/InvertedCaseWhen.java

```diff
@@ -0,0 +1,415 @@
+                ScalarOperator child1 = call.getChild(1);
+                if (!child0.isConstantRef()) {
+                    return visit(call, context);
+                }
+                CaseWhenOperator caseWhen =
+                        new CaseWhenOperator(call.getType(), null, child0,
+                                Lists.newArrayList(BinaryPredicateOperator.eq(child1, child0),
+                                        ConstantOperator.createNull(call.getType())));
+                return caseWhen.accept(this, context);
+            }
+            return visit(call, context);
+        }
+    }
+
+    private static Optional<ScalarOperator> or(List<ScalarOperator> args) {
+        if (args.isEmpty()) {
+            return Optional.empty();
+        } else if (args.size() == 1) {
+            return Optional.of(args.get(0));
+        } else {
+            return Optional.of(CompoundPredicateOperator.or(args));
+        }
+    }
+
+    private static final InvertCaseWhenVisitor INVERT_CASE_WHEN_VISITOR = new InvertCaseWhenVisitor();
+
+    public static Optional<InvertedCaseWhen> from(ScalarOperator op) {
+        return op.accept(INVERT_CASE_WHEN_VISITOR, null);
+    }
+
+    private static ScalarOperator ifThenNullOrFalse(ScalarOperator p) {
+        Function ifFunc = Expr.getBuiltinFunction(FunctionSet.IF, new Type[] {Type.BOOLEAN, Type.BOOLEAN, Type.BOOLEAN},
+                Function.CompareMode.IS_IDENTICAL);
+        Preconditions.checkArgument(ifFunc != null);
+        return new CallOperator(FunctionSet.IF, Type.BOOLEAN,
+                Lists.newArrayList(p, ConstantOperator.NULL, ConstantOperator.FALSE), ifFunc);
+    }
+
+    private static class SimplifyVisitor extends ScalarOperatorVisitor<Optional<ScalarOperator>, Void> {
+
+        @Override
+        public Optional<ScalarOperator> visit(ScalarOperator scalarOperator, Void context) {
+            return Optional.empty();
+        }
+
+        @Override
+        public Optional<ScalarOperator> visitInPredicate(InPredicateOperator predicate, Void context) {
+            if (predicate.getChildren().stream().skip(1)
+                    .allMatch(child -> !child.isConstantRef() || child.isConstantNull())) {
+                return Optional.empty();
+            }
+            Optional<InvertedCaseWhen> maybeInvertedCaseWhen = from(predicate.getChild(0));
+            if (!maybeInvertedCaseWhen.isPresent()) {
+                return Optional.empty();
+            }
+            InvertedCaseWhen invertedCaseWhen = maybeInvertedCaseWhen.get();
+            Set<ScalarOperator> inSet = predicate.getChildren().stream().skip(1).collect(Collectors.toSet());
+            boolean isNotIn = predicate.isNotIn();
+            Map<ScalarOperator, WhenAndOrdinal> selected = Maps.newHashMap();
+            Map<ScalarOperator, WhenAndOrdinal> unSelected = Maps.newHashMap();
+            invertedCaseWhen.thenMap.entrySet().forEach(e -> {
```

Comment:
Should we limit the number of child in the in list?

Reply:
see the code as follows,  case-when-without-case-clause and case-when-with-case-clause have different policies,  the comment explains in detail.